File: java/src/main/java/io/github/muthuishere/toolnexus/Adapters.java
package io.github.muthuishere.toolnexus;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Provider adapters — turn the uniform tool list into each LLM's tool schema.
 * Schema only; execution is identical for every provider.
 */
public final class Adapters {
    private Adapters() {}

    public static List<Map<String, Object>> toOpenAI(List<Tool> tools) {
        List<Map<String, Object>> out = new ArrayList<>();
        for (Tool t : tools) {
            Map<String, Object> fn = new LinkedHashMap<>();
            fn.put("name", t.name());
            fn.put("description", t.description());
            fn.put("parameters", t.inputSchema());
            Map<String, Object> wrapper = new LinkedHashMap<>();
            wrapper.put("type", "function");
            wrapper.put("function", fn);
            out.add(wrapper);
        }
        return out;
    }

    public static List<Map<String, Object>> toAnthropic(List<Tool> tools) {
        List<Map<String, Object>> out = new ArrayList<>();
        for (Tool t : tools) {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("name", t.name());
            m.put("description", t.description());
            m.put("input_schema", t.inputSchema());
            out.add(m);
        }
        return out;
    }

    public static List<Map<String, Object>> toGemini(List<Tool> tools) {
        List<Map<String, Object>> decls = new ArrayList<>();
        for (Tool t : tools) {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("name", t.name());
            m.put("description", t.description());
            m.put("parameters", t.inputSchema());
            decls.add(m);
        }
        Map<String, Object> wrapper = new LinkedHashMap<>();
        wrapper.put("functionDeclarations", decls);
        List<Map<String, Object>> out = new ArrayList<>();
        out.add(wrapper);
        return out;
    }
}

File: csharp/src/Toolnexus/Adapters.cs
namespace Toolnexus;

/// <summary>
/// Provider adapters — turn the uniform tool list into each LLM's tool schema.
/// Schema only; execution is identical for every provider.
/// </summary>
public static class Adapters
{
    public static List<Dictionary<string, object?>> ToOpenAI(IEnumerable<ITool> tools)
    {
        var output = new List<Dictionary<string, object?>>();
        foreach (var t in tools)
        {
            var fn = new Dictionary<string, object?>
            {
                ["name"] = t.Name,
                ["description"] = t.Description,
                ["parameters"] = t.InputSchema,
            };
            output.Add(new Dictionary<string, object?>
            {
                ["type"] = "function",
                ["function"] = fn,
            });
        }
        return output;
    }

    public static List<Dictionary<string, object?>> ToAnthropic(IEnumerable<ITool> tools)
    {
        var output = new List<Dictionary<string, object?>>();
        foreach (var t in tools)
        {
            output.Add(new Dictionary<string, object?>
            {
                ["name"] = t.Name,
                ["description"] = t.Description,
                ["input_schema"] = t.InputSchema,
            });
        }
        return output;
    }

    public static List<Dictionary<string, object?>> ToGemini(IEnumerable<ITool> tools)
    {
        var decls = new List<object?>();
        foreach (var t in tools)
        {
            decls.Add(new Dictionary<string, object?>
            {
                ["name"] = t.Name,
                ["description"] = t.Description,
                ["parameters"] = t.InputSchema,
            });
        }
        return new List<Dictionary<string, object?>>
        {
            new() { ["functionDeclarations"] = decls },
        };
    }
}
